### PR TITLE
JDK-8247589: Implementation of Alpine Linux/x64 Port

### PR DESCRIPTION
**Pull request by voitylov**
continuing the review thread from here https://mail.openjdk.java.net/pipermail/core-libs-dev/2020-September/068546.html

> The download side of using JNI in these tests is that it complicates the 
> setup a bit for those that run jtreg directly and/or just build the JDK 
> and not the test libraries. You could reduce this burden a bit by 
> limiting the load library/isMusl check to Linux only, meaning isMusl 
> would not be called on other platforms.
>
> The alternative you suggest above might indeed be better. I assume you 
> don't mean splitting the tests but rather just adding a second @test 
> description so that the vm.musl case runs the test with a system 
> property that allows the test know the expected load library path behavior.

I have updated the PR to split the two tests in multiple @test s. 

> The updated comment in java_md.c in this looks good. A minor comment on 
> Platform.isBusybox is Files.isSymbolicLink returning true implies that 
> the link exists so no need to check for exists too. Also the 
> if-then-else style for the new class in ProcessBuilder/Basic.java is 
> inconsistent with the rest of the test so it stands out.

Thank you, these changes are done in the updated PR.

> Given the repo transition this weekend then I assume you'll create a PR 
> for the final review at least. Also I see JEP 386 hasn't been targeted 
> yet but I assume Boris, as owner, will propose-to-target and wait for it 
> to be targeted before it is integrated.

Yes. How can this be best accomplished with the new git workflow?
- we can continue the review process till the end and I will request the integration to happen only after the JEP is targeted. I guess this step is now done by typing "slash integrate" in a comment.
- we can pause the review process now until the JEP is targeted.

In the first case I'm kindly asking the Reviewers who already chimed in on that to re-confirm the review here.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8247589](https://bugs.openjdk.java.net/browse/JDK-8247589): Implementation of Alpine Linux/x64 Port

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to f61f546a12c8fb0e81415240b6c080d6c5d04553
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to f61f546a12c8fb0e81415240b6c080d6c5d04553
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Contributors
 * Mikael Vidstedt `<mikael@openjdk.org>`
 * Alexander Scherbatiy `<alexsch@openjdk.org>`
 * Axel Siebenborn `<asiebenborn@openjdk.org>`
 * Aleksei Voitylov `<avoitylov@openjdk.org>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/49/head:pull/49`
`$ git checkout pull/49`

### CODE REVIEW

**Comment by AlanBateman** [comment]
This change was in review on core-libs-dev and other mailing lists before the switch to skara/git. The issues that I brought up have been added in the PR and I don't have any further comments.

**Review by AlanBateman — APPROVED** [review]

**Review by erikj79 — APPROVED** [review]
Build changes look ok.

File: test/hotspot/jtreg/runtime/StackGuardPages/exeinvoke.c [hunk 1]
```diff
@@ -258,14 +274,20 @@ int main (int argc, const char** argv) {
     exit(7);
   }
 
+  size_t stack_size = get_java_stacksize();
   pthread_t thr;
+  pthread_attr_t thread_attr;
+
+  pthread_attr_init(&thread_attr);
+  pthread_attr_setstacksize(&thread_attr, stack_size);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/StackGuardPages/exeinvoke.c:282]: Just a comment in response to the explanation as to why this change is needed. If the default thread stacksize under musl is insufficient to successfully attach such a thread to the VM then this will cause problems for applications that embed the VM directly (or which otherwise directly attach existing threads).

File: test/hotspot/jtreg/runtime/TLS/exestack-tls.c [hunk 2]
```diff
@@ -54,6 +57,7 @@ JNIEnv* create_vm(JavaVM **jvm, char* argTLS) {
     return env;
 }
 
+#if defined(__GLIBC)
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/TLS/exestack-tls.c:60]: Why do we use this form here but at line 30 we have:
#ifdef __GLIBC__
?

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Attempting to use the GitHub UI for fuirther review. If this doesn't work out well I will revert to direct email.

File: make/autoconf/platform.m4 [hunk 3]
```diff
@@ -493,6 +533,13 @@ AC_DEFUN([PLATFORM_SETUP_LEGACY_VARS_HELPER],
   fi
   AC_SUBST(HOTSPOT_$1_CPU_DEFINE)
 
+  if test "x$OPENJDK_$1_LIBC" = "xmusl"; then
```
[the remark this answers is not in the document]
Comment (voitylov) [reply make/autoconf/platform.m4]: this check is removed in the updated version. As a consequence, LIBC variable is added to the release file for all platforms, which is probably a good thing.

File: src/hotspot/os/linux/os_linux.cpp [hunk 4]
```diff
@@ -608,17 +618,24 @@ void os::Linux::libpthread_init() {
   #error "glibc too old (< 2.3.2)"
 #endif
 
+#ifdef MUSL_LIBC
+  // confstr() from musl libc returns EINVAL for
+  // _CS_GNU_LIBC_VERSION and _CS_GNU_LIBPTHREAD_VERSION
+  os::Linux::set_libc_version("unknown");
```
[the remark this answers is not in the document]
Comment (voitylov) [reply src/hotspot/os/linux/os_linux.cpp]: Right, this should be more consistent with glibc which here returns name and version. Updated as suggested.

File: src/hotspot/os/linux/os_linux.cpp [hunk 5]
```diff
@@ -608,17 +618,24 @@ void os::Linux::libpthread_init() {
...
+  os::Linux::set_libpthread_version("unknown");
```
[the remark this answers is not in the document]
Comment (voitylov) [reply src/hotspot/os/linux/os_linux.cpp]: The pthread version is also updated to "musl - unknown". Reason being, pthread functionality for musl is built into the library.

File: src/hotspot/share/runtime/abstract_vm_version.cpp [hunk 6]
```diff
@@ -257,12 +257,18 @@ const char* Abstract_VM_Version::internal_vm_info_string() {
     #define FLOAT_ARCH_STR XSTR(FLOAT_ARCH)
   #endif
 
+  #ifdef MUSL_LIBC
+    #define LIBC_STR "-" XSTR(LIBC)
+  #else
+    #define LIBC_STR ""
```
[the remark this answers is not in the document]
Comment (voitylov) [reply src/hotspot/share/runtime/abstract_vm_version.cpp:263]: Unlike the case above, I think it's best to keep it as is. I'd expect there to be a bunch of scripts in the wild which parse it and may get broken when facing a triplet for existing platforms.

File: src/jdk.hotspot.agent/linux/native/libsaproc/ps_proc.c [hunk 7]
```diff
@@ -279,8 +279,10 @@ static attach_state_t ptrace_attach(pid_t pid, char* err_buf, size_t err_buf_len
         return ATTACH_THREAD_DEAD;
       }
     }
-    char buf[200];
-    char* msg = strerror_r(errno, buf, sizeof(buf));
+    // To improve portability across platforms and avoid conflicts
+    // between GNU and XSI versions of strerror_r, plain strerror is used.
+    // It's safe because this code is not used in any multithreaded environment.
```
[the remark this answers is not in the document]
Comment (voitylov) [reply src/jdk.hotspot.agent/linux/native/libsaproc/ps_proc.c]: You are right! The updated version #ifdefs the XSI or GNU versions of strerror_r in this place. Note to self: file a bug to address the usage of strerror in other places, at least for HotSpot.

File: test/hotspot/jtreg/runtime/StackGuardPages/exeinvoke.c (hunk 1 above)
Comment (voitylov) [reply test/hotspot/jtreg/runtime/StackGuardPages/exeinvoke.c:282]: This fix https://git.musl-libc.org/cgit/musl/commit/src/aio/aio.c?id=1a6d6f131bd60ec2a858b34100049f0c042089f2  addresses the problem for recent versions of musl. The test passes on a recent Alpine Linux 3.11.6 (musl 1.1.24) and fails on Alpine Linux 3.8.2 (musl 1.1.19) without this test fix.

There are still older versions of the library in the wild, hence the test fix. The mitigation for such users would be a distro upgrade.

File: test/hotspot/jtreg/runtime/TLS/exestack-tls.c (hunk 2 above)
Comment (voitylov) [reply test/hotspot/jtreg/runtime/TLS/exestack-tls.c]: Fixed to be consistent.

File: test/hotspot/jtreg/runtime/StackGuardPages/exeinvoke.c (hunk 1 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/StackGuardPages/exeinvoke.c:282]: Thanks for the additional info on this.

**Comment by dholmes-ora** [comment]
Updates look good. Nothing further from me.

**Review by dholmes-ora — APPROVED** [review]

**Comment by voitylov** [comment]
thank you Alan, Erik, and David! When the JEP becomes Targeted, I'll use this PR to integrate the changes.

**Comment by voitylov** [comment]
I added the contributors that could be found in the portola project commits. If anyone knows some other contributors I missed, I'll be happy to stand corrected.

**Comment by dholmes-ora** [comment]
@voitylov For future reference please don't force-push commits on open PRs as it breaks the commit history. I can no longer just look at the two most recent commits and see what they added relative to what I had previously reviewed. Thanks.

**Comment by voitylov** [comment]
@dholmes-ora yes, sorry about that. I updated the branch to pull the recent changes to enable pre-submit tests and ensure everything is well before integration and though the branch looked good it confused the pull request. So I had to force push it back to the original state.

**Comment by voitylov** [comment]
@iignatev I resolved the conflict in whitebox.cpp and fixed a minor style nit on the way. Could you take a look?

**Comment by iignatev** [comment]
> @iignatev I resolved the conflict in whitebox.cpp and fixed a minor style nit on the way. Could you take a look?

LGTM

**Review by dholmes-ora — APPROVED** [review]

**Comment by voitylov** [comment]
Thanks everyone!
